core/src/SYCL/Kokkos_SYCL_Instance.cpp

```diff
@@ -55,6 +55,8 @@ namespace Impl {
 
 int SYCLInternal::was_finalized = 0;
 
+std::vector<std::optional<sycl::queue>*> SYCLInternal::all_queues;
```

Comment:
Why not just a `vector<queue>`?

`m_queue` is an `optional<queue>` to get around vendor bugs when assigning one `queue` to another.  As long as we aren't inserting/emplacing/erasing from the beginning or middle of the `vector` or assigning to elements in the `vector`, we won't run into that issue, and it won't be a concern at all once the vendor bugs are fixed.

Also, is `m_queue` going to eventually go away (I assume so but an checking here)?

Reply:
I was running into the same (or what looked like the same) vendor bug and storing pointers seemed to be the only way to make it work, see [`db9dc39` (#3767)](https://github.com/kokkos/kokkos/pull/3767/commits/db9dc3912cef79727324e7ae181ac31cc95a59fb) for the fixing commit and https://cloud.cees.ornl.gov/jenkins-ci/blue/organizations/jenkins/Kokkos/detail/Kokkos/4186/pipeline for the error when just using `std::vector<sycl::queue>`.

> Also, is m_queue going to eventually go away (I assume so but an checking here)? (Even now it could go away as it should always be equivalent to all_queues.back().

What do you mean? We still must have access to the `queue` corresponding to the current instance so I don't see a way to eliminate it. Or am I misunderstanding?

Reply:
As I read lines 119 and 120, m_queue and all_queues.back() always refer to the same queue, don't they?

Reply:
> As I read lines 119 and 120, m_queue and all_queues.back() always refer to the same queue, don't they?

That's only true for the last instance that adds a `queue` to the `static` variable.

Reply:
Please confirm that `std::vector<std::optional<sycl::queue>>` does not work

Reply:
Comment to capture this discussion